Update IntelliSense docs to mention indexing and auto imports behaviours and settings (#5637)
microsoft/vscode-docs: docs/python/settings-reference.md

| --- | --- | --- |
| typeCheckingMode | off | Specifies the level of type checking analysis to perform. Available values are `off`, `basic`, and `strict`. When set to `off` no type checking analysis is conducted; unresolved imports/variables diagnostics are produced. When set to `basic` non-type checking-related rules (all rules in `off`), as well as basic type checking rules are used. When set to `strict` all type checking rules at the highest severity of error (including all rules in `off` and `basic` categories) are used. |
| diagnosticMode | openFilesOnly | Specifies what code files the language server analyzes for problems. Available values are `workspace` and `openFilesOnly`. |

> | include | [] | Paths of directories or files that should be included in analysis. If no paths are specified, Pylance defaults to the directory that contains workspace root. Paths may contain wildcard characters such as `**` (a directory or multiple levels of directories), `*` (a sequence of zero or more characters), or `?` (a single character). |

Reviewer: ```suggestion
| include | [] | Paths of directories or files that should be included in analysis. If no paths are specified, Pylance defaults to the directory that contains the workspace root. Paths may contain wildcard characters such as `**` (a directory or multiple levels of directories), `*` (a sequence of zero or more characters), or `?` (a single character). |
```

> | include | [] | Paths of directories or files that should be included in analysis. If no paths are specified, Pylance defaults to the directory that contains the workspace root. Paths may contain wildcard characters such as `**` (a directory or multiple levels of directories), `*` (a sequence of zero or more characters), or `?` (a single character). |

Reviewer 2: You might consider breaking up this text with `<br>` tags where you think there is a logical break in the sentences.

```md
Paths of directories or files that should be included in analysis.<br>If no paths are specified, Pylance defaults to the directory that contains the workspace root. <br>Paths may contain wildcard characters such as `**` (a directory or multiple levels of directories), `*` (a sequence of zero or more characters), or `?` (a single character).
```

| exclude | [] | Paths of directories or files that should not be included in analysis. These override the directories listed under the `python.analysis.include` setting, allowing specific subdirectories to be excluded. Note that files listed in this `exclude` setting may still be included in the analysis if they are referenced/imported by source files that are not in the excluded list. Paths may contain wildcard characters such as `**` (a directory or multiple levels of directories), `*` (a sequence of zero or more characters), or `?` (a single character). If no exclude paths are specified, Pylance automatically excludes the following: `**/node_modules`, `**/\_\_pycache\_\_`, `.git` and any virtual environment directories. |

> | ignore | [] | Paths of directories or files which diagnostic output (errors and warnings) should be suppressed, even if they are an included file or within the transitive closure of an included file. Paths may contain wildcard characters such as `**` (a directory or multiple levels of directories), `*` (a sequence of zero or more characters), or `?` (a single character). If no value is provided, the value of `python.linting.ignorePatterns` (if set) will be used. |

Reviewer: ```suggestion
| ignore | [] | Paths of directories or files whose diagnostic output (errors and warnings) should be suppressed, even if they are an included file or within the transitive closure of an included file. Paths may contain wildcard characters such as `**` (a directory or multiple levels of directories), `*` (a sequence of zero or more characters), or `?` (a single character). If no value is provided, the value of `python.linting.ignorePatterns` (if set) will be used. |
```

| stubPath | ./typings | Specifies a path to a directory that contains custom type stubs. Each package's type stub file(s) are expected to be in its own subdirectory. |
| autoSearchPaths | true | Indicates whether to automatically add search paths based on some predefined names (like `src`). Available values are `true` and `false`. |
| extraPaths | [] | Specifies extra search paths for import resolution. Accepts paths specified as strings and separated by commas if there are multiple paths. For example: `["path 1","path 2"]` |

> | indexing | true | Used to specify whether Pylance should index user files as well as installed third party libraries at start up, to provide a more complete set of symbols in features such as auto-imports, quick fixes, auto completions, etc. Accepted values are `true` or `false`. When set to `true`, by default Pylance indexes top-level symbols of installed packages (i.e., symbols in `__all__` under `package/__init__.py`), as well as all symbols from up to 2000 user files. When set to `false`, Pylance will only display symbols already referenced or used in files that were previously opened in the editor.    |

Reviewer: Is it "were previously open" or "are currently open"? I'm not sure.

Reviewer 3: Thanks

Reviewer 4: when indexing is off, it is a bit more complex (or simpler based on how you see it I guess). when indexing is off, we use the original pyright's auto import behavior (I didn't change it since I didn't want to change pyright's behavior - indexing only exists in pylance, and not pyright). and the original pyright auto import behavior is simply if a file (module info) exists in memory, pyright offer auto import from module level symbols that exist in the memory. it doesn't care whether the file is user or library file. if it exists in memory, it will offer auto import. and it doesn't care why the file exists in memory. if it exists, it gather all module level symbols in the file, and offer them.

now, how the file can bring into the memory. there are many ways for files to load into memory.

1. a file is opened.
2. a file is used for import resolution. ex) a file is opened, and it has import x, and x module (file x.py) has import y. then file y.py will be loaded into memory. basically all files transitively used by a file opened will be loaded into the memory. this includes both user and library files.
3. a file is used for a IDE feature. ex) features such as rename, find all reference or call hierarchy require loading bunch of files to see whether they have references for the target symbol. all files checked and files transitively related will be loaded into the memory. this also mean as we change/add IDE feature it can affect auto import behavior by affecting what files are loaded into memory.
4. file once loaded into the memory won't be unloaded even if a file is closed.
5. files (module info) are only unloaded when pyright gets memory pressure. in that case, all loaded files will be dumped (emptied) and files related to currently opened files will be reloaded.
6. if a file is deleted from disk, then info for that file will be unloaded.
7. it doesn't care the file loaded is py or pyi. if it has module info, then it is good. so you can get auto import suggestion for packages not installed but has stubs in typeshed or bundled.

this is what I can remember on top of my head. but, basically, default pyright auto import behavior (when index is off) completely depends on what files are currently loaded. so, what user get could be different time to time based on current context and release.

when indexing is on, default behavior is what you described.

1. for user file, we index module level symbols up to 2000 user files. user symbol that are related to open files (modules that currently exist memory, see the above list on how one can be brought into the memory) will be offered in auto import otherwise, it only offered in code action (add import).
2. for library files, we index symbols in ```__all__``` from top level ```__init__``` file from installed packages. (no auto import if the package is not installed even if pyi exists) and offered in auto import and add import.

Reviewer 4: not sure how much detail needed to be documented.

Author: I was going to add more details, but I think it ended up being too much info to be digestible. I'm updating "were opened in or loaded by the editor" to be generic enough. If folks are interested in understanding the details they can reach out to us. It should also be fair for us to have a wiki page with more technical details if these sort of questions become more often.

> | indexing | true | Used to specify whether Pylance should index user files as well as installed third party libraries at start up, to provide a more complete set of symbols in features such as auto-imports, quick fixes, auto completions, etc. Accepted values are `true` or `false`. When set to `true`, by default Pylance indexes top-level symbols of installed packages (i.e., symbols in `__all__` under `package/__init__.py`), as well as all symbols from up to 2000 user files. When set to `false`, Pylance will only display symbols already referenced or used in files that were previously opened in or loaded by the editor.    |

Reviewer 2: "such as auto imports, Quick Fixes, ...." to match the text above

> | packageIndexDepths | [] | Used to override how many levels under installed packages to index on a per package basis. By default, only top-level modules are indexed (depth = 1). To index submodules, increase depth by 1 for each level of submodule you want to index. Accepts values are tuples of [```package name```, ```depth to scan```, ```include all symbols?```]. If `include all symbols?` is set to `false`, only symbols in each package's `__all__` are included. When it's set to `true`, Pylance will index every module/top level symbol declarations in the file. Usage example: ```[["pandas", 3, true], ["numpy", 2, false]] ``` |

Reviewer: ```suggestion
| packageIndexDepths | [] | Used to override how many levels under installed packages to index on a per package basis. By default, only top-level modules are indexed (depth = 1). To index submodules, increase depth by 1 for each level of submodule you want to index. Accepted values are tuples of [```package name```, ```depth to scan```, ```include all symbols?```]. If `include all symbols?` is set to `false`, only symbols in each package's `__all__` are included. When it's set to `true`, Pylance will index every module/top level symbol declarations in the file. Usage example: ```[["pandas", 3, true], ["numpy", 2, false]] ``` |
```
